Fix DQM_L1TMonitor_L1TStage2EMTF: there are four RPC planes in the Endcaps

perrotta opened the pull request:
#### PR description:

Fixes #37686

Since beginning of Run2 there are four RPC planes in the endcaps, but apparently L1TStage2EMTF was never updated for it.
The code for that class is rather involuted, and I cannot  be 100% sure that I fixed it everywhere: @cms-sw/rpc-dpg-l2 is expected to have a look and either approve this fix, or use it as a base for a more correct one (in a separate PR: I will close this one if so)

By the way, if really the error is there since the beginning of Run2, I wonder whether the DQM outputs of this code were ever used, or inspected, since then. Perhaps @cms-sw/rpc-dpg-l2 can decide that a more convenient fix could be simply to remove it from CMSSW as a whole, thus avoiding having to maintain some code which is not needed any more

#### PR validation:

It builds
`runTheMatrix.py -l 136.787` ends without errors

I did not check with UBSAN, because this error is apparently not listed any more in the UBSAN IB builds, see e.g. https://cmssdt.cern.ch/SDT/html/cmssdt-ib/#/relVal/CMSSW_12_4/2022-05-06-1100?selectedArchs=slc7_amd64_gcc11&selectedFlavors=UBSAN_X&selectedStatus=failed

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37853/29785

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @perrotta (Andrea Perrotta) for master.

It involves the following packages:

- DQM/L1TMonitor (**dqm**)


@emanueleusai, @ahmad3213, @cmsbuild, @jfernan2, @pmandrik, @micsucmed, @rvenditti can you please review it and eventually sign? Thanks.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37853/29786

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #37853 was updated. @emanueleusai, @ahmad3213, @cmsbuild, @jfernan2, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37853/29787

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #37853 was updated. @emanueleusai, @ahmad3213, @cmsbuild, @jfernan2, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0f6f4d/24515/summary.html
**COMMIT**: 7ca73c2579e0c8d3376556123ecc55205486f5e5
**CMSSW**: CMSSW_12_4_X_2022-05-06-2300/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/37853/24515/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3701451
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 2
 - DQMHistoTests: Total successes: 3701419
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 15756.759999999997 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 10024.0,... ): 787.838 KiB      L1T/L1TStage2EMTF
 - Checked 206 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-05-06-2300+0f6f4d/50109/triggerResults): no differences found

emanueleusai commented:
+1 from DQM side, but RPC DPG should follow up on Andrea's request before this is merged

emanueleusai commented:
hold

cmsbuild commented:
Pull request has been put on hold by  @emanueleusai
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

perrotta commented:
type bugfix

perrotta commented:
> * DQMHistoSizes: Histogram memory added: 15756.759999999997 KiB( 48 files compared)
> * DQMHistoSizes: changed ( 10024.0,... ): 787.838 KiB      L1T/L1TStage2EMTF
> * Checked 206 log files, 45 edm output root files, 49 DQM output files

To whoever will review this PR: please validate also the content of the newly added DQM histos for RPC

perrotta commented:
@cms-sw/rpc-dpg-l2 please let us know when do you expect to review this PR (just to be sure that someone in the RPC DPG is actually reached by these messages, and can take action for them)

emanueleusai commented:
- emailed conveners directly

perrotta commented:
> Hi @perrotta The current available version is correct. There are 4 RPC endcap stations and every of them has 2 rings. But not all of them go to EMTF. RE+-13, RE+-23 go to OMTF. So, I think there is no need of this change. Ping Karol in case he has another opinion. @kbunkow .

Thank you @mileva 

I think you refer to the updates I made to [lines 96-97](https://github.com/cms-sw/cmssw/pull/37853/files#diff-94bef1d606a7d3275d32be162fb054dda5a8ca65cda274c20a065d35e5fdd29eR96-R97) , where I added the RE+-13 and RE+-23 to the `rpc_name` and `rpc_label` arrays: this can probably get reverted.

However, the bug fix for the issue reported in https://github.com/cms-sw/cmssw/issues/37686 (which is the main reason for this PR) is at [line 88](https://github.com/cms-sw/cmssw/pull/37853/files#diff-c88b879f7901371cbd194cd5bf17e62abdf04f1066f15a63729eb1a1984685a4R88): if the `emtfTrackModeVsRPCBXDiff` has dimension 6 it cannot account for the fourth station, thus giving rise to the runtime error reported in https://github.com/cms-sw/cmssw/issues/37686

Moreover, there was another rather obvious bug at [line 978](https://github.com/cms-sw/cmssw/pull/37853/files#diff-94bef1d606a7d3275d32be162fb054dda5a8ca65cda274c20a065d35e5fdd29eR978) : e.g. station 1 would get the same `iRPC` in the two endcaps

mileva commented:
Thank you for the details, @perrotta !
I agree with you.

> I think you refer to the updates I made to [lines 96-97]
Yes, I refereed to these lines, exactly.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37853/30014

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #37853 was updated. @emanueleusai, @ahmad3213, @cmsbuild, @jfernan2, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

perrotta commented:
Following https://github.com/cms-sw/cmssw/pull/37853#issuecomment-1125340719 and the discussions before it, I reverted the extention of the EMTF monitoring to the whole RPC (RE+-13, RE+-23 go to OMTF, not to EMTF).
Only the two bug fixes mentioned in https://github.com/cms-sw/cmssw/pull/37853#issuecomment-1125259564 remain.

perrotta commented:
Following https://github.com/cms-sw/cmssw/pull/37853#issuecomment-1125340719 and the discussions before it, I reverted the extention of the EMTF monitoring to the whole RPC (RE+-13, RE+-23 go to OMTF, not to EMTF).
Only the two bug fixes mentioned in https://github.com/cms-sw/cmssw/pull/37853#issuecomment-1125259564 remain.

perrotta commented:
please test

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0f6f4d/24723/summary.html
**COMMIT**: 57b9cbbb30fd4e4fb4763c82caf76028a8b2c3de
**CMSSW**: CMSSW_12_4_X_2022-05-13-2300/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37853/24723/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3697377
 - DQMHistoTests: Total failures: 86
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3697269
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 32.999999999999986 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 10024.0,... ): 1.650 KiB        L1T/L1TStage2EMTF
 - Checked 206 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-05-13-2300+0f6f4d/50295/triggerResults): no differences found

emanueleusai commented:
Are the differences in Tracking DQM plots of workflow 12434.0 caused by the bugfix?

https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf12434_0_pr/CMSSW_12_4_X-PRcmssw_37853-50295/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf12434_0_base/CMSSW_12_4_X-PRcmssw_37853-50295/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;

The 2 ME added also appear to be empty in all workflows. I assume this is caused by the small statistics of the samples?

emanueleusai commented:
unhold

perrotta commented:
> Are the differences in Tracking DQM plots of workflow 12434.0 caused by the bugfix?

 Unlikely...
Let retry the tests

> The 2 ME added also appear to be empty in all workflows. I assume this is caused by the small statistics of the samples?

This must be checked before merging: perhaps there is some issue with how those histos are filled.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0f6f4d/24724/summary.html
**COMMIT**: 57b9cbbb30fd4e4fb4763c82caf76028a8b2c3de
**CMSSW**: CMSSW_12_4_X_2022-05-14-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37853/24724/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3697377
 - DQMHistoTests: Total failures: 97
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3697257
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 33.00399999999999 KiB( 48 files compared)
 - DQMHistoSizes: changed ( 10024.0,... ): 1.650 KiB        L1T/L1TStage2EMTF
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 206 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-05-14-1100+0f6f4d/50296/triggerResults): no differences found

perrotta commented:
For 12434.0 the difference is in the Tracking DQM, where there is one less track which had a largely displaced vertex
![image](https://user-images.githubusercontent.com/4069749/168555859-567094b9-4ef3-4179-be34-c22561cdee9c.png)
Of course, this PR which only touches the DQM monitoring for L1TStage2EMTF can hardly be tought responsible for it. But let keep investigating the possible origin.

perrotta commented:
More related to the RPC: @mileva @kbunkow @cms-sw/rpc-dpg-l2 could you please investigate (perhaps with some larger statistic than the validation comparisons that are automatically performed by bot. that the content of the `emtfTrackModeVsRPCBXDiff` plots does make sense now, also taking into account the comment of @emanueleusai  in  https://github.com/cms-sw/cmssw/pull/37853#issuecomment-1126808184 ?

eyigitba commented:
Hi @perrotta , I think the suggested change makes sense. These plots were added last year along with GEM timing plots for EMTF. Apprently we missed this bug. I see that the `labelRPCStation` also needs to be updated for this to work.

perrotta committed:
There are four RPC planes in the Endcaps

perrotta commented:
> Hi @perrotta , I think the suggested change makes sense. These plots were added last year along with GEM timing plots for EMTF. Apprently we missed this bug. I see that the `labelRPCStation` also needs to be updated for this to work.

Good catch, thank you @eyigitba. Also `nameRPCStation`.
Fixed, and squashed all commits into one

perrotta commented:
> Hi @perrotta , I think the suggested change makes sense. These plots were added last year along with GEM timing plots for EMTF. Apprently we missed this bug. I see that the `labelRPCStation` also needs to be updated for this to work.

Good catch, thank you @eyigitba. Also `nameRPCStation`.
Fixed, and squashed all commits into one

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37853/30030

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #37853 was updated. @emanueleusai, @ahmad3213, @cmsbuild, @jfernan2, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0f6f4d/24740/summary.html
**COMMIT**: 64ecb05cfa101bc94a111e8fc340b2e0acca4bee
**CMSSW**: CMSSW_12_4_X_2022-05-16-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/37853/24740/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-0f6f4d/11634.301_TTbar_14TeV+2021_Run3FS+TTbar_14TeV_TuneCP5_GenSim+HARVESTNano

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3741432
 - DQMHistoTests: Total failures: 91
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3741318
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 33.084 KiB( 49 files compared)
 - DQMHistoSizes: changed ( 10024.0,... ): 1.654 KiB        L1T/L1TStage2EMTF
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 208 log files, 45 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_4_X_2022-05-16-1100+0f6f4d/50312/triggerResults): no differences found

emanueleusai commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
